File: imperativo/usuariodb.cpp
#include <iostream>
#include <sqlite3.h>
#include <vector>
#import "constantes.cpp"
#import "utilitario.cpp"

using namespace std;

int autenticaUsuario(struct Usuario &usuario);
void criaTabelaUsuario();
int editaUsuario(struct Usuario &usuario);
int insereUsuario(struct Usuario &usuario);
int obtemID();
int removeUsuario(int id);

/**
 * Autentica o usuario atraves do email e senha.
 *
 * @param usuario
 * @return 0 (sucesso) e 1 (erro)
 */
int autenticaUsuario(struct Usuario &usuario) {
    criaTabelaUsuario();

    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao autenticar usuario: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Nao foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "SELECT id, nome, ficcao, nao_ficcao, romance, horror, biografia "
                 "FROM usuario "
                 "WHERE email = '" + usuario.email + "' AND senha = '" + usuario.senha + "' "
                 "LIMIT 1;";

    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE)
            break;

        if (retorno != SQLITE_ROW) {
            exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return 1;
        }

        usuario.id = sqlite3_column_int(stmt, 0);
        usuario.nome = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 1));
        usuario.ficcao = sqlite3_column_int(stmt, 2);
        usuario.naoFiccao = sqlite3_column_int(stmt, 3);
        usuario.romance = sqlite3_column_int(stmt, 4);
        usuario.horror = sqlite3_column_int(stmt, 5);
        usuario.biografia = sqlite3_column_int(stmt, 6);
    }

    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return 0;
}

void criaTabelaUsuario() {
    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao criar a tabela de usuario: ";

    if (retorno) {
        cerr << "Não foi possível abrir o banco de dados: " << sqlite3_errmsg(bancoDados) << endl;
    }

    string sql = "CREATE TABLE IF NOT EXISTS usuario("
                 "id INT PRIMARY KEY NOT NULL, "
                 "nome TEXT NOT NULL, "
                 "email TEXT NOT NULL UNIQUE, "
                 "senha INT NOT NULL, "
                 "ficcao INT, "
                 "nao_ficcao INT, "
                 "romance INT, "
                 "horror INT, "
                 "biografia INT);";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        cerr << mensagemErro << sqlite3_errmsg(bancoDados) << endl;
        sqlite3_free(erroBanco);
    }

    sqlite3_close(bancoDados);
}

/**
 * Edita as informacoes de usuario.
 *
 * @param usuario
 * @return 0 (sucesso) e 1 (erro)
 */
int editaUsuario(struct Usuario &usuario) {
    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao editar as informaçéos do usuário: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "UPDATE usuario "
                 "SET nome = '" + usuario.nome + "', "
                     "email = '" + usuario.email + "', "
                     "senha = '" + usuario.senha + "', "
                     "ficcao = '" + to_string(usuario.ficcao) + "', "
                     "nao_ficcao = '" + to_string(usuario.naoFiccao) + "', "
                     "romance = '" + to_string(usuario.romance) + "', "
                     "horror = '" + to_string(usuario.horror) + "', "
                     "biografia = '" + to_string(usuario.biografia) + "'"
                 "WHERE id = " + to_string(usuario.id) + ";";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_free(erroBanco);
        sqlite3_close(bancoDados);

        return 1;
    }

    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Consulta um usuario.
 * @param id
 * @return 0 (sucesso) e 1 (erro)
 */
int consultaUsuario(vector<string> &usuario, int id) {
    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao consultar usuário: ";

    if (retorno != SQLITE_OK) {
        cerr << "Não foi possível abrir o banco de dados: " << sqlite3_errmsg(bancoDados) << endl;
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "SELECT * FROM usuario WHERE id = " + to_string(id) + ";";
    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        cerr << mensagemErro << sqlite3_errmsg(bancoDados) << endl;
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE) break;

        if (retorno != SQLITE_ROW) {
            cerr << mensagemErro << sqlite3_errmsg(bancoDados) << endl;
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return 1;
        }

        const char *idUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_ID));
        const char *nomeUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_NOME));
        const char *emailUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_EMAIL));
        const char *interessesUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_INTERESSES));

        usuario.push_back(idUsuario);
        usuario.push_back(nomeUsuario);
        usuario.push_back(emailUsuario);
        usuario.push_back(interessesUsuario);
    }

    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Insere as informacoes de usuario.
 *
 * @param usuario
 * @return 0 (sucesso) e 1 (erro)
 */
int insereUsuario(struct Usuario &usuario) {
    criaTabelaUsuario();
    usuario.id = obtemID();

    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao inserir usuário: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "INSERT INTO usuario(id, nome, email, senha, ficcao, nao_ficcao, romance, horror, biografia) "
                 "VALUES('" + to_string(usuario.id) + "', '" + usuario.nome + "', '" + usuario.email + "', '" +
                 usuario.senha + "', '" + to_string(usuario.ficcao) + "', '" + to_string(usuario.naoFiccao) + "', '" +
                 to_string(usuario.romance) + "', '" + to_string(usuario.horror) + "', '" +
                 to_string(usuario.biografia) + "');";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_free(erroBanco);
        sqlite3_close(bancoDados);

        return 1;
    }

    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Consulta um usuario.
 * @param id
 * @return ultimo id (sucesso) e -1 (erro)
 */
int obtemID() {
    int id = 1;
    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao consultar usuário: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return -1;
    }

    string sql = "SELECT id FROM usuario ORDER BY id DESC LIMIT 1;";
    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return -1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE) break;

        if (retorno != SQLITE_ROW) {
            exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return -1;
        }
        id = sqlite3_column_int(stmt, U_ID);
        id += 1;
    }
    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return id;
}

/**
 * Remove o usuario pelo id.
 *
 * @param id
 * @return 0 (sucesso) e 1 (erro)
 */
int removeUsuario(int id) {
    sqlite3 *bancoDados;
    char *erroBanco;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao remover usuário: ";

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados("Não foi possível abrir o banco de dados: ", sqlite3_errmsg(bancoDados));
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "DELETE FROM usuario "
                 "WHERE id = " + to_string(id) + ";";

    retorno = sqlite3_exec(bancoDados, sql.c_str(), NULL, 0, &erroBanco);

    if (retorno != SQLITE_OK) {
        exibeMensagemErroBancoDados(mensagemErro, sqlite3_errmsg(bancoDados));
        sqlite3_free(erroBanco);
        sqlite3_close(bancoDados);

        return 1;
    }

    sqlite3_close(bancoDados);

    return 0;
}

/**
 * Lista todos os usuarios.
 * @param usuarios
 * @return 0 (sucesso) e 1 (erro)
 */
int listaTodosUsuario(vector<vector<string> > &usuarios) {
    sqlite3 *bancoDados;
    sqlite3_stmt *stmt;
    int retorno = sqlite3_open(BANCO_DADOS, &bancoDados);
    string mensagemErro = "Ocorreu um erro ao listar usuário: ";

    if (retorno != SQLITE_OK) {
        cerr << "Não foi possível abrir o banco de dados: " << sqlite3_errmsg(bancoDados) << endl;
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    string sql = "SELECT * FROM usuario;";
    retorno = sqlite3_prepare(bancoDados, sql.c_str(), -1, &stmt, NULL);

    if (retorno != SQLITE_OK) {
        cerr << mensagemErro << sqlite3_errmsg(bancoDados) << endl;
        sqlite3_finalize(stmt);
        sqlite3_close(bancoDados);

        return 1;
    }

    while (true) {
        retorno = sqlite3_step(stmt);

        if (retorno == SQLITE_DONE) break;

        if (retorno != SQLITE_ROW) {
            cerr << mensagemErro << sqlite3_errmsg(bancoDados) << endl;
            sqlite3_finalize(stmt);
            sqlite3_close(bancoDados);

            return 1;
        }

        const char *idUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_ID));
        const char *nomeUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_NOME));
        const char *emailUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_EMAIL));
        const char *interessesUsuario = reinterpret_cast<const char *>(sqlite3_column_text(stmt, U_INTERESSES));

        vector<string> usuario;
        usuario.push_back(idUsuario);
        usuario.push_back(nomeUsuario);
        usuario.push_back(emailUsuario);
        usuario.push_back(interessesUsuario);

        usuarios.push_back(usuario);
    }

    sqlite3_finalize(stmt);
    sqlite3_close(bancoDados);

    return 0;
}

File: funcional/src/UsuarioDB.hs
module UsuarioDB where

import           Database.HDBC
import           Database.HDBC.Sqlite3
import           EstanteType

-- Autentica usuario atraves do email e senha.
autenticaUsuario :: String -> String -> IO Usuario
autenticaUsuario email senha = do
  conexao <- connectSqlite3 "prompskoob.db"
  resultado <- quickQuery' conexao "SELECT * FROM usuario WHERE email = ? AND senha = ?;" [toSql email, toSql senha]
  disconnect conexao
  if null resultado
    then return (obtemUsuarioNulo)
    else return (preencheUsuario (head resultado))

-- Cria uma tabela de usuario.
criaTabelaUsuario :: IO ()
criaTabelaUsuario = do
  conexao <- connectSqlite3 "prompskoob.db"
  run
    conexao
    ("CREATE TABLE IF NOT EXISTS usuario(id INTEGER PRIMARY KEY NOT NULL, nome TEXT NOT NULL, email TEXT NOT NULL " ++
     "UNIQUE, senha TEXT NOT NULL, ficcao INT DEFAULT (0), nao_ficcao INT DEFAULT (0), romance INT DEFAULT (0), " ++
     "horror INT DEFAULT (0), biografia INT DEFAULT (0));")
    []
  commit conexao
  disconnect conexao

-- Remove o usuario do sistema.
removeUsuario :: Int -> IO Bool
removeUsuario idUsuario = do
  conexao <- connectSqlite3 "prompskoob.db"
  removido <- run conexao "DELETE FROM usuario WHERE id = ?;" [toSql idUsuario]
  commit conexao
  disconnect conexao
  return (removido >= 1)

-- Verifica se o e-mail ja foi cadastrado no banco de dados
verificaSeEmailExiste :: String -> IO Bool
verificaSeEmailExiste email = do
  conexao <- connectSqlite3 "prompskoob.db"
  resultado <- quickQuery' conexao "SELECT id FROM usuario WHERE email = ?;" [toSql email]
  disconnect conexao
  return (size resultado > 0)

-- Insere um novo usuario no sistema.
insereUsuario :: Usuario -> IO Bool
insereUsuario usuario = do
  conexao <- connectSqlite3 "prompskoob.db"
  inserido <-
    run
      conexao
      ("INSERT INTO usuario(nome, email, senha, ficcao, nao_ficcao, romance, horror, biografia) VALUES " ++
       "(?, ?, ?, ?, ?, ?, ?, ?);")
      [ toSql (nome usuario)
      , toSql (email usuario)
      , toSql (senha usuario)
      , toSql (gostaFiccao usuario)
      , toSql (gostaNaoFiccao usuario)
      , toSql (gostaRomance usuario)
      , toSql (gostaHorror usuario)
      , toSql (gostaBiografia usuario)
      ]
  commit conexao
  disconnect conexao
  return (inserido >= 1)

consultaUsuario :: Int -> IO Usuario
consultaUsuario codigo = do
  conexao <- connectSqlite3 "prompskoob.db"
  resultado <- quickQuery' conexao "SELECT * FROM usuario WHERE id = ? LIMIT 1;" [toSql codigo]
  disconnect conexao
  return (preencheUsuario (head resultado))

editaUsuario :: Usuario -> IO Usuario
editaUsuario usuario = do
  conexao <- connectSqlite3 "prompskoob.db"
  run
    conexao
    ("UPDATE usuario SET nome = ?, email = ?, senha = ?, ficcao = ?, nao_ficcao = ?, romance = ?, horror = ?, " ++
     "biografia = ? WHERE id = ?;")
    [ toSql (nome usuario)
    , toSql (email usuario)
    , toSql (senha usuario)
    , toSql (gostaFiccao usuario)
    , toSql (gostaNaoFiccao usuario)
    , toSql (gostaRomance usuario)
    , toSql (gostaHorror usuario)
    , toSql (gostaBiografia usuario)
    , toSql (idUsuario usuario)
    ]
  commit conexao
  disconnect conexao
  usuarioEditado <- consultaUsuario (idUsuario usuario)
  return (usuarioEditado)

preencheUsuario :: [SqlValue] -> Usuario
preencheUsuario [sqlId, sqlNome, sqlEmail, sqlSenha, sqlFiccao, sqlNaoFiccao, sqlRomance, sqlHorror, sqlBiografia] =
  Usuario
    { idUsuario = codigoSql
    , nome = nomeSql
    , email = emailSql
    , senha = senhaSql
    , gostaFiccao = ficcaoSql
    , gostaNaoFiccao = naoFiccaoSql
    , gostaRomance = romanceSql
    , gostaHorror = horrorSql
    , gostaBiografia = biografiaSql
    , estantes = []
    }
  where
    codigoSql = fromSql sqlId :: Int
    nomeSql = fromSql sqlNome :: String
    emailSql = fromSql sqlEmail :: String
    senhaSql = fromSql sqlSenha :: String
    ficcaoSql = fromSql sqlFiccao :: Int
    naoFiccaoSql = fromSql sqlNaoFiccao :: Int
    romanceSql = fromSql sqlRomance :: Int
    horrorSql = fromSql sqlHorror :: Int
    biografiaSql = fromSql sqlBiografia :: Int

obtemUsuarioNulo :: Usuario
obtemUsuarioNulo = Usuario 0 "NULL" "NULL" "NULL" 0 0 0 0 0 []

isUsuarioNulo :: Usuario -> Bool
isUsuarioNulo usuario
  | (idUsuario usuario == 0) && (nome usuario == "NULL") = True
  | otherwise = False
